File: daily_python/590_N_ary_Tree_Postorder_Traversal.py
import os
import sys
from typing import List, Optional
import queue

# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children

def create_tree(nodes: List[int])-> Optional['Node']:
    if not nodes:
        return None
    n = len(nodes)
    root = Node(nodes[0])
    index = 2
    que = queue.Queue()
    que.put(root)
    while index < n and not que.empty():
        temp = que.get()
        _children = []
        while index < n and nodes[index] != -1:
            _children.append(Node(nodes[index]))
            que.put(_children[-1])
            index += 1
        temp.children = _children
        index += 1
    return root

class Solution:
    def traversal(self, node: 'Node')-> None:
        if not node:
            return
        if node.children:
            for child in node.children:
                self.traversal(child)
        self.path.append(node.val)
    def postorder(self, root: 'Node') -> List[int]:
        self.path = []
        self.traversal(root)
        return self.path
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nodes = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            dataset.append(nodes)
    return dataset

if __name__ == '__main__':
    if (len(sys.argv) == 1):
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for nodes in dataset:
        node = create_tree(nodes)
        results.append(solution.postorder(node))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')

File: daily/2024/590_N_ary_Tree_Postorder_Traversal.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};

Node* create_tree(vector<int> nodes){
    if (nodes.empty()) return nullptr;
    int n = nodes.size();
    Node* root = new Node(nodes[0]);
    queue<Node*> que;
    que.push(root);
    int index = 2;
    while(index < n){
        Node* temp = que.front();
        que.pop();
        vector<Node*> _children;
        while(index < n && nodes[index] != -1){
            _children.push_back(new Node(nodes[index]));
            que.push(_children.back());
            index++;
        }
        temp->children = _children;
        index++;
    }
    return root;
}

class Solution {
public:
    vector<int> postorder_path;
    void traversal(Node* root){
        if (!root) return;
        for (Node* n : root->children) traversal(n);
        postorder_path.push_back(root->val);
    }
    vector<int> postorder(Node* root) {
        traversal(root);
        return postorder_path;
    }
};

int main(){
    vector<int> nodes = {1,-1,3,2,4,-1,5,6};
    // vector<int> nodes = {1,-1,2,3,4,5,-1,-1,6,7,-1,8,-1,9,10,-1,-1,11,-1,12,-1,13,-1,-1,14};
    Node* root = create_tree(nodes);
    Solution S;
    vector<int> ans = S.postorder(root);
    for (int i : ans) cout << i << " ";
    cout << endl;
    return 0;
}